File: Thread/app/src/main/java/com/nipun/thread/MainActivity.java
package com.nipun.thread;

import androidx.appcompat.app.AppCompatActivity;
import androidx.cardview.widget.CardView;

import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.airbnb.lottie.LottieAnimationView;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();
    private Button mBtnStart1;
    private Button mBtnStart2;
    private LottieAnimationView p1;
    private LottieAnimationView p2;
    private BackGroundTask looperThread;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mBtnStart1 = findViewById(R.id.btn1);
        mBtnStart2 = findViewById(R.id.btn2);
        p1 = findViewById(R.id.bar1);
        p2 = findViewById(R.id.bar2);
        looperThread = new BackGroundTask();

        mBtnStart1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                for (int i = 0; i < 10; i++) {
                    Log.d("Nipun", i + "");
                    SystemClock.sleep(1000);
                }
                Toast.makeText(MainActivity.this, "Handler 1 start", Toast.LENGTH_SHORT).show();

                Thread thread = new Thread();
                thread.start();
                Toast.makeText(MainActivity.this, "Handler 1 end", Toast.LENGTH_SHORT).show();
                p1.setVisibility(View.VISIBLE);
                p2.setVisibility(View.GONE);
            }
        });

        mBtnStart2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                for (int i = 0; i < 10; i++) {
                    Log.d("Nipun", i + "");
                    SystemClock.sleep(1000);
                }
                looperThread.start();
                Toast.makeText(MainActivity.this, "Handler 2 end", Toast.LENGTH_SHORT).show();
                p2.setVisibility(View.VISIBLE);
                p1.setVisibility(View.GONE);
            }
        });

    }
}

File: IntentService/app/src/main/java/com/nipun/intentservice/MainActivity.kt
package com.nipun.intentservice

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.nipun.intentservice.BoundService.MyBinder
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {


    var mBoundService: BoundService? = null
    var mServiceBound = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        bind.setOnClickListener { view ->
            val intent = Intent(applicationContext, BoundService::class.java)
            bindService(intent, mServiceConnection, Context.BIND_AUTO_CREATE)
            Log.d("Nipun", "Bind")
            Toast.makeText(this, "Service is bind", Toast.LENGTH_SHORT).show()

        }
        stop_service.setOnClickListener { view ->
            if (mServiceBound) {
                unbindService(mServiceConnection);
                mServiceBound = false;
                Log.d("Nipun", "UnBind")
                Toast.makeText(this, "Service is Unbind", Toast.LENGTH_SHORT).show()
            }


        }
        print_Detail.setOnClickListener { view ->
            if (mServiceBound) {
                getName.setText(mBoundService?.getName())
                getAge.setText(mBoundService?.getAge())
                getAddress.setText(mBoundService?.getAddress())
                Log.d("Nipun", "PrintDetail")
                Toast.makeText(this, "Printing Detail", Toast.LENGTH_SHORT).show()

            } else {
                Toast.makeText(this, "Please Bind your service first", Toast.LENGTH_SHORT).show()
            }

        }

    }

    private val mServiceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(name: ComponentName) {
            mServiceBound = false
        }

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            val myBinder = service as MyBinder
            mBoundService = myBinder.service
            mServiceBound = true
        }
    }

    override fun onStart() {
        super.onStart()
    }

    override fun onStop() {
        super.onStop()
        if (mServiceBound) {
            unbindService(mServiceConnection)
            mServiceBound = false
        }
    }
}